File: LeetCode/Medium/0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> merged = new ArrayList<>();
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);

        for (int[] i : intervals) {
            if (!merged.isEmpty() && i[0] <= merged.get(merged.size() - 1)[1]) {
                merged.get(merged.size() - 1)[1] = Math.max(merged.get(merged.size() - 1)[1], i[1]);
            } else {
                merged.add(i);
            }
        }

        return merged.stream().toArray(int[][]::new);
    }
}

File: LeetCode/Medium/0056-merge-intervals/0056-merge-intervals.kt
class Solution {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        val merged = mutableListOf<IntArray>()
        intervals.sortBy { it[0] }

        for (i in intervals) {
            if (merged.isNotEmpty() && i[0] <= merged.last()[1]) {
                merged.last()[1] = maxOf(merged.last()[1], i[1])
            } else {
                merged.add(i)
            }
        }

        return merged.toTypedArray()
    }
}